packages/runtime-common/realm.ts

```diff
@@ -526,6 +551,223 @@ export class Realm {
+        errors.push({
+          title,
+          detail,
+          status: 422,
+        });
+      }
+    }
+    return errors;
+  }
+
+  // this method carefully checks before writing with the intent of
+  // stopping failed operations that depend on others
+  private async checkBeforeAtomicWrite(
+    operations: AtomicOperation[],
+  ): Promise<ErrorDetails[]> {
+    let promises = [];
+    for (let { href } of operations) {
+      promises.push(this.#adapter.exists(href));
+    }
+    let booleanFlags = await Promise.all(promises);
+    return operations
+      .filter((_, i) => booleanFlags[i])
+      .map(({ href }) => {
+        return {
+          title: 'Resource already exists',
+          detail: `Resource ${href} already exists`,
+          status: 409,
+        };
+      });
+  }
+
+  private lowestStatusCode(errors: ErrorDetails[]): number {
+    let statuses = errors
+      .filter((e) => e.status)
+      .map((e) => e.status) as number[];
+    return Math.min(...statuses);
+  }
+
+  private async handleAtomicOperations(
+    request: Request,
+    requestContext: RequestContext,
+  ): Promise<Response> {
+    let body = await request.text();
+    let json;
+    try {
+      json = JSON.parse(body);
+    } catch (e) {
+      return createResponse({
+        body: JSON.stringify({
+          errors: [
+            {
+              title: 'Invalid atomic:operations format',
+              detail: `Request body is not valid JSON`,
+            },
+          ],
+        }),
+        init: {
+          status: 400,
+          headers: {
+            'content-type': SupportedMimeType.CardJson,
+          },
+        },
+        requestContext,
+      });
+    }
+    let validationErrors = this.validate(json);
+    if (validationErrors.length > 0) {
+      return createResponse({
+        body: JSON.stringify({ errors: validationErrors }),
+        init: { status: 400, headers: { 'content-type': 'application/json' } }, //consolidate to 400
+        requestContext,
+      });
+    }
+    let operations = json['atomic:operations'] as AtomicOperation[];
+
+    let atomicCheckErrors = await this.checkBeforeAtomicWrite(operations);
+    if (atomicCheckErrors.length > 0) {
+      return createResponse({
+        body: JSON.stringify({ errors: atomicCheckErrors }),
+        init: {
+          status: this.lowestStatusCode(atomicCheckErrors),
+          headers: { 'content-type': 'application/json' },
+        },
+        requestContext,
+      });
+    }
+    let sourceFiles = new Map<LocalPath, string>();
+    let instanceFiles = new Map<LocalPath, string>();
+    let sourceOperations = filterOperationsWithSourceData(operations);
+    let instanceOperations = filterOperationsWithCardData(operations);
+    let instanceWriteResults: WriteResult[] = [];
+    let sourceWriteResults: WriteResult[] = [];
+
+    // we intentionally write source files first so that fileSerialization can work
```

Comment:
Just to point out. In the case, if an instance depends on a module in its adoptsFrom, fileSerialization tries to fetch the module somehow
- Previously, we thought we should allow the that fetch to go directly to the file system rather than the index (thats its default behaviour) by maybe including a header 
- I found that `getFileWithFallbacks` still relies on the file being written first. So, what I did, is I perform a write of the sources first 
- if we want better implemnentation for batch writes, we should introduce a new `bulkWrite` api that the realm can use